### ISSUE TITLE
Don't close completion window on pressed modifier.

### ISSUE THREAD

**Issue by Janfel** [howl-editor/howl#520]
Currently the autocompletion window closes when a modifier key (such as `Ctrl`, `Alt`, `ISO_Level5_Shift`) is pressed. I am using a keymap, that gives me arrow keys on `Mod5+e s d f`, but I can not use them, because the window closes when pressing `Mod5`.

**Comment by Janfel** [comment]
The issue seems to be caused by [this line][line]. Implementing a more advanced check would propably fix it.

[line]: https://github.com/howl-editor/howl/blob/ad89244f32060c7532c25226fdf60bf3f4b0d15f/lib/howl/ui/completion_popup.moon#L34

**Comment by Janfel** [comment]
I'm sorry, it's actually [this line][line]. The name of the function seemed like it was from a string library.

[line]: https://github.com/howl-editor/howl/blob/ad89244f32060c7532c25226fdf60bf3f4b0d15f/lib/howl/ui/completion_popup.moon#L9

**Comment by shalabhc** [comment]
How about we only close the completion on `escape`?

**Comment by Janfel** [comment]
That sounds reasonable. Maybe add `Home` and `End` to that list.

**Comment by shalabhc** [comment]
So I think this happens in menu_popup.moon:`on_unhandled`. Interestingly it uses a custom keymap to navigate the menu but `ctrl_p` and `ctrl_n` in there will never work because `ctrl` first triggers the `on_unhandled` which closes the menu popup.

What keymap are you using to map the Mod combinations to arrows? How is this configured? It may still not work unless we put a `binding_for` in the menupopup kemap.

**Comment by shalabhc** [comment]
A simple fix here would be to just ignore plain `ctrl`, `alt` and `shift` key presses in the `on_unhandled` function above. This would still close the popup on any other non character keypress (`escape`, `home` etc.) and would allow firing the modified key combinations (e.g. `ctrl_p`). However if you bound some other keystroke to 'cursor-down', I don't think that will work. That will have to be bound via `binding_for` ([indirect bindings](https://howl.io/versions/0.6/doc/api/bindings.html#indirect-bindings)).

**Comment by Janfel** [comment]
I am using the standard German/Neo 2 layout (`de_neo`) that comes with Arch/Manjaro. The keys are implemented directly in the X keymap. Here is the output of `xev`
```
KeyPress event, serial 28, synthetic NO, window 0x9e00001,
    root 0x1d7, subw 0x0, time 274940317, (583,129), root:(3374,726),
    state 0x0, keycode 108 (keysym 0xfe11, ISO_Level5_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 94
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 28, synthetic NO, window 0x9e00001,
    root 0x1d7, subw 0x0, time 274940761, (583,129), root:(3374,726),
    state 0x20, keycode 26 (keysym 0xff52, Up), same_screen YES,
    XKeysymToKeycode returns keycode: 111
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 28, synthetic NO, window 0x9e00001,
    root 0x1d7, subw 0x0, time 274940889, (583,129), root:(3374,726),
    state 0x20, keycode 26 (keysym 0xff52, Up), same_screen YES,
    XKeysymToKeycode returns keycode: 111
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 28, synthetic NO, window 0x9e00001,
    root 0x1d7, subw 0x0, time 274941175, (583,129), root:(3374,726),
    state 0x20, keycode 39 (keysym 0xff51, Left), same_screen YES,
    XKeysymToKeycode returns keycode: 113
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 28, synthetic NO, window 0x9e00001,
    root 0x1d7, subw 0x0, time 274941287, (583,129), root:(3374,726),
    state 0x20, keycode 39 (keysym 0xff51, Left), same_screen YES,
    XKeysymToKeycode returns keycode: 113
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 28, synthetic NO, window 0x9e00001,
    root 0x1d7, subw 0x0, time 274941512, (583,129), root:(3374,726),
    state 0x20, keycode 40 (keysym 0xff54, Down), same_screen YES,
    XKeysymToKeycode returns keycode: 116
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 28, synthetic NO, window 0x9e00001,
    root 0x1d7, subw 0x0, time 274941604, (583,129), root:(3374,726),
    state 0x20, keycode 40 (keysym 0xff54, Down), same_screen YES,
    XKeysymToKeycode returns keycode: 116
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 28, synthetic NO, window 0x9e00001,
    root 0x1d7, subw 0x0, time 274941866, (583,129), root:(3374,726),
    state 0x20, keycode 41 (keysym 0xff53, Right), same_screen YES,
    XKeysymToKeycode returns keycode: 114
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 28, synthetic NO, window 0x9e00001,
    root 0x1d7, subw 0x0, time 274941961, (583,129), root:(3374,726),
    state 0x20, keycode 41 (keysym 0xff53, Right), same_screen YES,
    XKeysymToKeycode returns keycode: 114
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 28, synthetic NO, window 0x9e00001,
    root 0x1d7, subw 0x0, time 274942198, (583,129), root:(3374,726),
    state 0x20, keycode 108 (keysym 0xfe13, ISO_Level5_Lock), same_screen YES,
    XKeysymToKeycode returns keycode: 23
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False
```

**Comment by shalabhc** [comment]
@Janfel does #534 fix it for you?

**Comment by Janfel** [comment]
It does not. It would however, if you added `iso_level5_shift` and, for completeness sake, `iso_level3_shift` to the table.

**Comment by shalabhc** [comment]
Oh do these show up as modifier keys? Would be useful if you can paste the output of `describe-key` for the keypress that shouldn't trigger the pop-up close. I'm not really sure how to translate the `xev` output to Howl's key translations.

**Comment by Janfel** [comment]
Here you go:
```lua
-- Key translations (usable from bindings):
{
  "iso_level5_shift",
  "65041"
}

-- Key event:
{
  alt = false,
  control = false,
  key_code = 65041,
  key_name = "iso_level5_shift",
  lock = false,
  meta = false,
  shift = false,
  super = false
}

-- Bound command:
{}
```
and
```lua
-- Key translations (usable from bindings):
{
  "iso_level3_shift",
  "65027"
}

-- Key event:
{
  alt = false,
  control = false,
  key_code = 65027,
  key_name = "iso_level3_shift",
  lock = false,
  meta = false,
  shift = false,
  super = false
}

-- Bound command:
{}
```
I had to copy them using the primary clipboard. Is there a more intuitive solution?

**Comment by Janfel** [comment]
And to complete the list of modifier keys, here are `hyper_l` and `hyper_r`, in case someone decides to bind them.
```lua
-- Key translations (usable from bindings):
{
  "hyper_l",
  "65517"
}

-- Key event:
{
  alt = false,
  control = false,
  key_code = 65517,
  key_name = "hyper_l",
  lock = false,
  meta = false,
  shift = false,
  super = false
}
```
and
```lua
-- Key translations (usable from bindings):
{
  "hyper_r",
  "65518"
}

-- Key event:
{
  alt = false,
  control = false,
  key_code = 65518,
  key_name = "hyper_r",
  lock = false,
  meta = false,
  shift = false,
  super = false
}
```

**Comment by shalabhc** [comment]
Thanks. Now I'm wondering if there's a better way to handle these in bindings.moon. Shouldn't these be properly identified as modifier keys, similar to `ctrl`, `alt`, `super` etc?

**Comment by shalabhc** [comment]
Like if you press iso_level3_shift and some random character - is that translated properly in describe-key?

**Comment by Janfel** [comment]
I'm not entirely sure what you mean by "properly", but this is how they behave. `shift`+`l` produces `{"L", "shift_l"}` (excluding the keycodes), while `iso_level3_shift`+`l` produces just `{"up"}`, not `{"up", "iso_level5_shift_up"}`.

How about a table (maybe user-editable) listing all keys that Howl recognizes as modifiers, containing all of `alt`, `meta`, `shift`, `super`, `hyper`, `iso_level3_shift` and `iso_level5_shift`. This would ensure all modifiers are handled uniformly and make it easy to add further OS specific modifiers such as `command` or `windows_key` conditionally.

**Comment by shalabhc** [comment]
> while iso_level3_shift+l produces just {"up"}, not {"up", "iso_level5_shift_up"}.

I assume that's because it becomes 'up' as per the X keymap even before Howl sees the event. If this were not mapped, I would expect something like `{"l", "iso_level5_shift_up"}.

> How about a table (maybe user-editable) listing all keys that Howl recognizes as modifiers

Yeah this is what I was thinking. Seems like the right solution here. I wouldn't worry about user editable at this point.

**Comment by shalabhc** [comment]
OK a proper fix where we create composite translations like 'mod4_up' is a bit more involved and would possibly require modifying ljglibs/util.moon. Also I don't have a keyboard that has these keys so I can't really test it.

However I've made a shorter PR, similar to the previous one. See #540.

**Comment by shalabhc** [comment]
Fixed in #540
